### ISSUE TITLE
DbHelper.FromID not updated,  DbAccessViolation

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3432]
Hi!

I've written code that ensures that an instance of e.g. Person exists with the name Carl. Simplified steps:
1. Search for the instance on the current thread
2. If it doesn't exist switch to another attached thread and create the instance in Db.Transact.
3. Handle constraint violations and reread the instance from db if someone else created it at the same time.
4. Reload the instance on the main thread using DbHelper.FromID. The reason for reloading is that I got DbAccessViolation if using the created instance directly. So the main thread sometimes can't access and read the created and commited instance.

99.99% of the time this logic works just fine. But sometimes step4 fails. I'm using minimal transacts.
There seems to be a slight delay in the DbHelper.FromIDs index? Because if I uncomment: ```//NamedLog("Created OID: {0}", oid.Value);```
 before returning to the main thread in step4 it's 100% accurate.

It took me quite a while to reproduce my production code in a simplified example, still not simple...
```
using System;
using Starcounter;
using System.Text;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Threading;

namespace EnsureTest
{

    class Program
    {
        public static DbSession Singleton = new DbSession();
        private static bool s_failed;

        static void Main()
        {
            Item.CreateIndex();
            Item.DeleteAll();

            while (!s_failed)
            {
                Parallel.For(0, 100, CreateInstance); 
            }
        }

        private static void CreateInstance(int i)
        {
            if (s_failed)
            {
                Log("ParallelJob: {0}, skipped failed flag set", i);
                return;
            }
            Singleton.RunSync(() =>
                {
                    Log("ParallelJob: {0}", i);
                    // Should create the instance
                    var helper = new EnsureHelper("Carl-" + i + "-" + DateTime.Now);
                    Item item = helper.Ensure();
                    Log(helper.ToString());

                    // Should read the instance
                    item = helper.Ensure();
                    Log(helper.ToString());
                }, SchedulerPartitioner.GetNext());
        }

        private static void Log(string msg, params object[] args)
        {
            var timestamp = DateTime.Now.ToLongTimeString();
            var tid = "[" + Thread.CurrentThread.ManagedThreadId + "] ";
            var timeMsg = "[" + timestamp + "]" + tid + msg;
            Debug.Print(timeMsg, args);
            Console.WriteLine(timeMsg, args);
        }

        private class EnsureHelper
        {
            private readonly string m_name;
            private readonly string m_logPrefix;
            private bool m_isCreated;
            private Item m_instance;
            

            public EnsureHelper(string name)
            {
                m_name = name;
                m_logPrefix = string.Format("[{0}]", m_name);
            }

            private void NamedLog(string msg, params object[] args)
            {
                Log(m_logPrefix + msg, args);
            }

            public Item Ensure()
            {
                m_instance = null;
                m_isCreated = false;

                m_instance = ReadInstance();

                if (m_instance == null)
                {
                    m_instance = Create();
                }

                return m_instance;
            }

            private Item ReadInstance()
            {
                return Db.SQL<Item>("SELECT i FROM EnsureTest.Item i WHERE i.Name=?", m_name).First;
            }

            private Item Create()
            {
                Item instance = null;

                for (int i = 0; i < 100; i++)
                {
                    NamedLog("Create Attempt: {0}/{1}", i, 100);
                    try
                    {
                        // Executes in another thread and transact
                        ulong? oid = DoCreate();

                        if (oid.HasValue)
                        {
                            // Reload the instance using object id
                            instance = (Item)DbHelper.FromID(oid.Value);

                            if (instance != null)
                            {
                                m_isCreated = true;
                                break;
                            }
                            else
                            {
                                NamedLog("Failed to retrieve newly created instance: {0}", oid.Value);
                            }
                        }
                        // Try reading 
                        instance = ReadInstance();

                        if (instance != null)
                        {
                            break;
                        }
                    }
                    catch (Exception ex)
                    {
                        s_failed = true;
                        instance = ReadInstance();

                        if (instance != null)
                        {
                            break;
                        }
                        NamedLog("Error: " + ex);
                    }
                }
                return instance;
            }

            private ulong? DoCreate()
            {
                ulong? oid = null;

                var tid = Thread.CurrentThread.ManagedThreadId;
                // Run as task to change thread
                Task.Factory.StartNew(() =>
                    {
                        Singleton.RunSync(() =>
                            {
                                try
                                {
                                    Db.Transact(() =>
                                        {
                                            var tid2 = Thread.CurrentThread.ManagedThreadId;

                                            if (tid == tid2)
                                            {
                                                throw new ApplicationException("No running on new thread");
                                            }

                                            var instance = new Item(m_name);
                                            oid = instance.GetObjectNo();
                                        });
                                    // If I uncomment this log DbHelper will always find the instance on the other thread.
                                    // Why is that, some context switching that takes enough time for the DbHelper to be updated?
                                    //NamedLog("Created OID: {0}", oid.Value);
                                }
                                catch (Exception ex)
                                {
                                    s_failed = true;
                                    NamedLog("OID: {0}{1}{2}", oid.Value, Environment.NewLine, ex);
                                    oid = null;
                                }
                            });
                    }).Wait();

                return oid;
            }

            public override string ToString()
            {
                return string.Format("[{0}], OID: {1}, IsCreated:{2}", m_name, m_instance.GetObjectNo(), m_isCreated);
            }
        }


    }

    [Database]
    public class Item
    {
        public Item(string name)
        {
            Name = name;
        }

        public readonly string Name;

        internal static void CreateIndex()
        {
            try
            {
                using (var t = new Transaction(false))
                {
                    Db.SQL("CREATE UNIQUE INDEX uname ON EnsureTest.Item (Name ASC)");
                    t.Commit();
                }
            }
            catch
            {
                // If already created
            }
        }

        internal static void DeleteAll()
        {
            foreach (var item in Db.SQL<Item>("SELECT i FROM EnsureTest.Item i"))
            {
                Db.Transact(item.Delete);
            }
        }
    }

    /// <summary>
    /// Distributes over the schedulers using round robin
    /// </summary>
    public class SchedulerPartitioner
    {
        private static readonly byte s_maxId = Db.Environment.SchedulerCount;
        private static int s_schedulerId;

        public static byte GetNext()
        {
            int expected = 0, current = 0;
            do
            {
                expected = (byte)(s_schedulerId + 1);
                current = s_schedulerId;

                if (expected == s_maxId)
                {
                    expected = 0;
                }

            } while (Interlocked.CompareExchange(ref s_schedulerId, expected, current) != current);

            return (byte)expected;
        }
    }
}
```

**Comment by careri** [comment]
If I add a method that retries reloading the instance for 1 second. The instance is always returned, most of the times directly but sometimes with a delay of a few thousand ticks
```
// Reload the instance using object id, was DbHelper.FromID
instance = Reload(oid.Value);
```

The method:
```
public static readonly TimeSpan s_maxWait = TimeSpan.FromSeconds(1);
public static readonly TimeSpan s_sleep = TimeSpan.FromTicks(1);

private Item Reload(ulong oid)
{
    var instance = (Item)DbHelper.FromID(oid);

    if (instance == null)
    {
        var watch = Stopwatch.StartNew();
        TimeSpan ts = default(TimeSpan);

        while (instance == null &&
            watch.Elapsed < s_maxWait)
        {
            Thread.Sleep(s_sleep);
            ts = watch.Elapsed;
            instance = (Item)DbHelper.FromID(oid);
        }

        if (instance == null)
        {
            throw new ApplicationException(string.Format("Failed to reload: {0}, OID: {1}", typeof(Item).Name, oid));
        }
        else
        {
            NamedLog("OID: {0} found after: {1} ticks", typeof(Item).Name, oid, ts.Ticks);
        }
    }
    return instance;
}
```

**Comment by careri** [comment]
Actually I just got the Failed to reload exception now, so maybe 1sec isn't enough after all

**Comment by careri** [comment]
I guess this regards @eriohl or @per-samuelsson?

**Comment by eriohl** [comment]
What do you mean by DbAccessViolation? AccessViolationException from DbHelper.FromID? Which version are you using?

**Comment by careri** [comment]
Version: 2.1.1327

I was getting AccessViolationException if I used the instance directly without reloading it with DbHelper.FromID.

So I tried to solve the problem by reloading it with DbHelper.FromID, since I was certain that the object was written. But to my surprise DbHelper.FromID returns null.

If I use the Reload-method described above the instance will be available after a while, usually just a couple of ms, but sometimes it takes longer than a second.

**Comment by eriohl** [comment]
Ok. A transaction runs in a snapshot so unless the transaction was created after a record was created it generally won't see that record. There is some exceptions to this because the snapshot of a transaction can be refreshed. I would guess without knowing that the reason for you not seeing the record is because it has not been. Someone else can answer this better.

Regarding the access violation: Do you have a stack trace for it?

**Comment by careri** [comment]
Can I force a refresh of the snapshot?

I'll try to achieve the stack trace again.

Very simplified this is what I do:
```
using Starcounter;
using System;
using System.Threading;

namespace CreateOnOtherThread
{
    class Program
    {
        private static readonly DbSession Singleton = new DbSession();

        static void Main()
        {
            // Run other thread
            var outerThread = new Thread(OuterTask);
            outerThread.Name = "outer";
            outerThread.Start();
        }

        private static void OuterTask(object o)
        {
            Singleton.RunSync(() =>
            {
                var proxy = new Proxy();
                var innerThread = new Thread(InnerTask);
                innerThread.Name = "inner";
                innerThread.Start(proxy);
                innerThread.Join();

                // This is where I get the exception. Access Violation
                try
                {
                    var instance = proxy.Instance;
                    var name = instance.Name;
                    Console.WriteLine(name);
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine(ex.ToString());
                }
            }, 0);
        }

        private static void InnerTask(object o)
        {
            var p = (Proxy)o;
            Singleton.RunSync(p.CreateItem);
        }

        class Proxy
        {
            public Item Instance { get; private set; }

            internal void CreateItem()
            {
                Db.Transact(() =>
                    {
                        Instance = new Item("New");
                    });
            }
        }
    }

    [Database]
    public class Item
    {
        public readonly string Name;

        public Item(string name)
        {
            Name = name;
        }
    }
}
```

**Comment by eriohl** [comment]
Ok. Thank you. Just to be clear: You are getting an managed AccessViolationException (that in this case would be intercepted on managed level). Right?

The only way I could see this happening is that kernel get string produces and error but the managed wrapper thinks the operation succeeded and tries to create a managed string from the (invalid) output.

**Comment by careri** [comment]
My bad, not an Access Violation. Got the exception a couple of hours before I rewrote the example and registered this issue so I mixed up the terms.

I get an ScErrInvalidObjectAccess on this row in the code below:
```var name = instance.Name;```

But the question persists, can I force a refresh of the snapshot? The behavior is inconsistent because 99.99% of the times I don't get the exception

This is the full code to reproduce:
```
using System;
using Starcounter;
using System.Text;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Threading;

namespace EnsureTest
{

    class Program
    {
        public static DbSession Singleton = new DbSession();
        private static bool s_failed;

        static void Main()
        {
            Item.CreateIndex();
            Item.DeleteAll();

            while (!s_failed)
            {
                Parallel.For(0, 100, CreateInstance); 
            }
        }

        private static void CreateInstance(int i)
        {
            if (s_failed)
            {
                Log("ParallelJob: {0}, skipped failed flag set", i);
                return;
            }
            Singleton.RunSync(() =>
                {
                    Log("ParallelJob: {0}", i);
                    // Should create the instance
                    var helper = new EnsureHelper("Carl-" + i + "-" + DateTime.Now);
                    Item item = helper.Ensure();
                    Log(helper.ToString());

                    // Should read the instance
                    item = helper.Ensure();
                    Log(helper.ToString());
                }, SchedulerPartitioner.GetNext());
        }

        private static void Log(string msg, params object[] args)
        {
            var timestamp = DateTime.Now.ToLongTimeString();
            var tid = "[" + Thread.CurrentThread.ManagedThreadId + "] ";
            var timeMsg = "[" + timestamp + "]" + tid + msg;
            Debug.Print(timeMsg, args);
            Console.WriteLine(timeMsg, args);
        }

        private class EnsureHelper
        {
            private readonly string m_name;
            private readonly string m_logPrefix;
            private bool m_isCreated;
            private Item m_instance;
            

            public EnsureHelper(string name)
            {
                m_name = name;
                m_logPrefix = string.Format("[{0}]", m_name);
            }

            private void NamedLog(string msg, params object[] args)
            {
                Log(m_logPrefix + msg, args);
            }

            public Item Ensure()
            {
                m_instance = null;
                m_isCreated = false;

                m_instance = ReadInstance();

                if (m_instance == null)
                {
                    m_instance = Create();
                }

                return m_instance;
            }

            private Item ReadInstance()
            {
                return Db.SQL<Item>("SELECT i FROM EnsureTest.Item i WHERE i.Name=?", m_name).First;
            }

            private Item Create()
            {
                Item instance = null;

                for (int i = 0; i < 100; i++)
                {
                    NamedLog("Create Attempt: {0}/{1}", i, 100);
                    try
                    {
                        // Executes in another thread and transact
                        var tuple = DoCreate();
                        var oid = tuple.Item1;
                        instance = tuple.Item2;

                        if (oid.HasValue)
                        {
                            try
                            {
                                var name = instance.Name;
                            }
                            catch (Exception ex)
                            {
                                NamedLog("Failed to read name from created instance: {0}", ex);
                                instance = null;
                            }

                            if (instance == null)
                            {
                                // Reload the instance using object id
                                instance = Reload(oid.Value);

                                if (instance != null)
                                {
                                    m_isCreated = true;
                                    break;
                                }
                                else
                                {
                                    NamedLog("Failed to retrieve newly created instance: {0}", oid.Value);
                                }
                            }
                        }
                        // Try reading 
                        instance = ReadInstance();

                        if (instance != null)
                        {
                            break;
                        }
                    }
                    catch (Exception ex)
                    {
                        s_failed = true;
                        instance = ReadInstance();

                        if (instance != null)
                        {
                            break;
                        }
                        NamedLog("Error: " + ex);
                    }
                }
                return instance;
            }

            private Tuple<ulong?,Item> DoCreate()
            {
                ulong? oid = null;
                Item instance = null;

                var tid = Thread.CurrentThread.ManagedThreadId;
                // Run as task to change thread
                Task.Factory.StartNew(() =>
                    {
                        Singleton.RunSync(() =>
                            {
                                try
                                {
                                    Db.Transact(() =>
                                        {
                                            var tid2 = Thread.CurrentThread.ManagedThreadId;

                                            if (tid == tid2)
                                            {
                                                throw new ApplicationException("No running on new thread");
                                            }

                                            instance = new Item(m_name);
                                            oid = instance.GetObjectNo();
                                        });
                                    // If I uncomment this log DbHelper will always find the instance on the other thread.
                                    // Why is that, some context switching that takes enough time for the DbHelper to be updated?
                                    //NamedLog("Created OID: {0}", oid.Value);
                                }
                                catch (Exception ex)
                                {
                                    s_failed = true;
                                    NamedLog("OID: {0}{1}{2}", oid.Value, Environment.NewLine, ex);
                                    oid = null;
                                    instance = null;
                                }
                            });
                    }).Wait();

                return Tuple.Create(oid, instance);
            }

            public static readonly TimeSpan s_maxWait = TimeSpan.FromSeconds(1);
            public static readonly TimeSpan s_sleep = TimeSpan.FromTicks(1);

            private Item Reload(ulong oid)
            {
                var instance = (Item)DbHelper.FromID(oid);

                if (instance == null)
                {
                    var watch = Stopwatch.StartNew();
                    TimeSpan ts = default(TimeSpan);

                    while (instance == null &&
                        watch.Elapsed < s_maxWait)
                    {
                        Thread.Sleep(s_sleep);
                        ts = watch.Elapsed;
                        instance = (Item)DbHelper.FromID(oid);
                    }

                    if (instance == null)
                    {
                        throw new ApplicationException(string.Format("Failed to reload: {0}, OID: {1}", typeof(Item).Name, oid));
                    }
                    else
                    {
                        NamedLog("OID: {0} found after: {1} ticks", typeof(Item).Name, oid, ts.Ticks);
                    }
                }
                return instance;
            }

            public override string ToString()
            {
                return string.Format("[{0}], OID: {1}, IsCreated:{2}", m_name, m_instance.GetObjectNo(), m_isCreated);
            }
        }


    }

    [Database]
    public class Item
    {
        public Item(string name)
        {
            Name = name;
        }

        public readonly string Name;

        internal static void CreateIndex()
        {
            try
            {
                using (var t = new Transaction(false))
                {
                    Db.SQL("CREATE UNIQUE INDEX uname ON EnsureTest.Item (Name ASC)");
                    t.Commit();
                }
            }
            catch
            {
                // If already created
            }
        }

        internal static void DeleteAll()
        {
            foreach (var item in Db.SQL<Item>("SELECT i FROM EnsureTest.Item i"))
            {
                Db.Transact(item.Delete);
            }
        }
    }

    /// <summary>
    /// Distributes over the schedulers using round robin
    /// </summary>
    public class SchedulerPartitioner
    {
        private static readonly byte s_maxId = Db.Environment.SchedulerCount;
        private static int s_schedulerId;

        public static byte GetNext()
        {
            int expected = 0, current = 0;
            do
            {
                expected = (byte)(s_schedulerId + 1);
                current = s_schedulerId;

                if (expected == s_maxId)
                {
                    expected = 0;
                }

            } while (Interlocked.CompareExchange(ref s_schedulerId, expected, current) != current);

            return (byte)expected;
        }
    }
}
```

**Comment by k-rus** [comment]
If I understand your code correctly the logic is in `Create`, which first calls `DoCreate` and then `Reload`. `DoCreate` might create an object in different thread then `Create` runs. Thus `Create` with `Reload` has own transaction scope (implicit), which might started before the transaction of creating the object. Thus the object is not available for `Reload` as `Reload` has older snapshot, i.e., before the creation of the object.

I am not sure but creating explicit transaction scope for `Reload` might help.

cc @chrhol

**Comment by eriohl** [comment]
> My bad, not an Access Violation. Got the exception a couple of hours before I rewrote the example and registered this issue so I mixed up the terms.

> I get an ScErrInvalidObjectAccess ...

Ok. Thank you for sorting that out. Then both problems are the same. ScErrInvalidObjectAccess you get when record with id doesn't exist in the view of the transaction. Same as when DbHelper.FromID returns null.

> But the question persists, can I force a refresh of the snapshot? The behavior is inconsistent because 99.99% of the times I don't get the exception.

As for this with the implicit transactions and transaction isolation I can't help you. I know how it works but not how it is supposed to be used. @per-samuelsson @chrhol @dan31?

But the simplest approach to solving the problem in this instance I think it so create a transaction after call to ```DoCreate``` (using ```Db.Transact``` as @k-rus suggests perhaps). But I'm guessing that this isn't the problem you actually want to solve.

**Comment by careri** [comment]
@k-rus There is not might in that. It is always created on another thread, and that is intentional. And as I state 99.99% of the times there is no problem transferring the instance between the transacts and therefore reload is never called.

But in some cases, I get the ScErrInvalidObjectAccess and therefore I call Reload. Reload doesn't switch transact, but after 1ms-10secs the instance is usually found. God knows why? But I actually had some cases where Reload fails as well, very uncommon though.

**Comment by careri** [comment]
@eriohl according to @chrhol calling Db.Transact if already in a Db.Transact has no effect at all. Only the "outer" Db.Transact does anything with transactions.

**Comment by careri** [comment]
Actually that last statement from me is why I change threads. I want to ensure that my object is fully commited and not meta-committed due to the fact that I might have a Db.Transact somewhere in the callstack

**Comment by eriohl** [comment]
> @eriohl according to @chrhol calling Db.Transact if already in a Db.Transact has no effect at all. Only the "outer" Db.Transact does anything with transactions.

True. But there is no outer Db.Transact in your case because you are using the implicit read-only transaction. This Db.Transact will create a new transaction.

**Comment by careri** [comment]
Well, I want to use the implicit read-only transactaction. I design my application to run on the implicit read-only transaction. Only switching to a Db.Transact when I do writes. The writes are really small tasks without any external IO (logging) as suggested by you @eriohl.

So I'm not quite following, is using the implicit read-only transaction a bad thing?

> Thus Db.Transact will create a new transaction

Well yes, and I guess the instane would be valid in that Transact. But the problem persists with my main code running on the implicit read-only transaction not being able to access the instance.

> But of course. This doesn't solve this problem.

What do you mean by this statement? Is there a flaw in my design?

**Comment by careri** [comment]
The thing is this 99.99% working, if it would be consistent not working I could live with it. 
But the exact same code running on multiple threads yielding different results is not something that feels stable to me.

Either this approach should be supported (I can't see a reason why not), or it shouldn't, Working most of the time is not something I can live with

**Comment by k-rus** [comment]
> The thing is this 99.99% working, if it would be consistent not working I could live with it. 
> But the exact same code running on multiple threads yielding different results is not something that feels stable to me.
> 
> Either this approach should be supported (I can't see a reason why not), or it shouldn't, Working most of the time is not something I can live with

You are running multi-threaded environment and your program does not guarantee consistent view.



> Well, I want to use the implicit read-only transactaction. I design my application to run on the implicit read-only transaction. Only switching to a Db.Transact when I do writes. The writes are really small tasks without any external IO (logging) as suggested by you @eriohl.
> 
> So I'm not quite following, is using the implicit read-only transaction a bad thing?

Implicit transaction is not a good thing. It is given to users for convenience and works enough well in simple scenarios. Since it is implicit, you do not control when it updates its snapshot of the database.

In your current example `Db.Transact` will work. I believe it is better to use `Db.Scope`, i.e., read-committed transactions, however I don't know how it works together with the implicit transaction.

You mentioned earlier that you actually do not need to run separate thread. Thus I am wonder why you are not running all your code in a single transaction. Using multiple threads without need is not good. Since one thread depends on another thread it reduces performance of your application.

**Comment by careri** [comment]
> You are running multi-threaded environment and your program does not guarantee consistent view.
Well, that sure isn't an explanation. The behaviour of starcounter is inconsistent, not my code in this case. If you could point out why most of the time the instance is valid and sometimes it not. How does the refresh of the implicit transaction occur?

Is there more information of this implicit readonly transaction? Is it always the same instance? Or is a new instance created for each call to DbSession.RunSync?

Multi-threaded environment isn't exactly an uncommon approach. I'm going to have hundreds of clients reporting status to my application. I can't very well run all of this on a single thread. The client's will share certain singleton instances. Made unique by index constraints.
As long as my clients only query data they do run in the same transaction, the readonly implict one.
According to @chrhol this was a good approach to writing an application. So this is quite confusing.

Long running transactions as you suggest isn't a good thing according to our experience at Heads. @eriohl also suggests keeping away from them since there is a limited number of fast slots for long running transactions as discussed in another issue.This goes for Db.Scope as well since it actually creates a transaction instead of using the lightweight Transact handles.

So you kinda contradict your self.

**Comment by eriohl** [comment]
> > > Actually that last statement from me is why I change threads. I want to ensure that my object is fully commited and not meta-committed due to the fact that I might have a Db.Transact somewhere in the callstack

> > But of course. This doesn't solve this problem.

> What do you mean by this statement? Is there a flaw in my design?

What I mean is that if you may have a Db.Transact in your callstack then Db.Transact won't create a new transaction. So it your design of rescheduling tasks on a different scheduler is because you might have a Db.Transact in your callstack then doing this won't work.

Whether there is a flaw in your design or not I can not advice on as mentioned. This with mixing long-running (including the implicit read-only transaction) and short running ones falls outside my area of expertise.

> Well, I want to use the implicit read-only transactaction. I design my application to run on the implicit read-only transaction. Only switching to a Db.Transact when I do writes. The writes are really small tasks without any external IO (logging) as suggested by you @eriohl.

I'm sorry but there seems to be some misunderstanding. I generally do not advice on application design because as mentioned: I don't know how this stuff is supposed to be used. Nor do I do so here as I tried to be clear about. Again sorry if there has been some misunderstanding on this. I only involved my self in this because you mentioned there was a access violation (which implies a kernel bug).

**Comment by careri** [comment]
We've hade similar disussions before: #3192

I tried adding a new readonly transaction with merging writes for each DbSession.RunSync like this:
```
Singleton.RunSync(() =>
{
    using (var t = new Transaction(true, false))
    {
```

But the problem persists, so this is not due to the implicit transaction. 
Second argument from the inline documentation:
> detectConflicts:
        //     Transaction conflicts detection flag (merging writes are used if False).

That tells me that the out transaction should read the instance created by the other thread right?

**Comment by k-rus** [comment]
> Well, that sure isn't an explanation. The behaviour of starcounter is inconsistent, not my code in this case. If you could point out why most of the time the instance is valid and sometimes it not.

Starcounter behaves according the specification. Read committed transaction such as implicit transaction or long running transaction reads committed values, but does not guarantee that all values, which were committed after the transaction was started, are visible. This is exactly your case.



> How does the refresh of the implicit transaction occur?

Not really correct question. I guess you wonder when committed data are available in read-committed transactions and I don't know.



> Is there more information of this implicit readonly transaction? Is it always the same instance? Or is a new instance created for each call to DbSession.RunSync?

I cannot answer this question as I don't know such details.



> Multi-threaded environment isn't exactly an uncommon approach. I'm going to have hundreds of clients reporting status to my application. I can't very well run all of this on a single thread. The client's will share certain singleton instances. Made unique by index constraints.
> As long as my clients only query data they do run in the same transaction, the readonly implict one.
> According to @chrhol this was a good approach to writing an application. So this is quite confusing.

I am speaking about exactly your case above. Creating another transaction and waiting for it to finish does not sound a good idea. 

In general multi-threaded is important to scale concurrency. When you design your application for multiple threads you should take in account that Starcounter provides snapshot isolation, not serializable isolation.



> Long running transactions as you suggest isn't a good thing according to our experience at Heads. @eriohl also suggests keeping away from them since there is a limited number of fast slots for long running transactions as discussed in another issue.This goes for Db.Scope as well since it actually creates a transaction instead of using the lightweight Transact handles.



> Long running transactions as you suggest isn't a good thing according to our experience at Heads. @eriohl also suggests keeping away from them since there is a limited number of fast slots for long running transactions as discussed in another issue.This goes for Db.Scope as well since it actually creates a transaction instead of using the lightweight Transact handles.
> 
> So you kinda contradict your self.

I don't contradict myself. I will clarify I suggest that you use `Db.Scope` instead of implicit read transaction. Using `Db.Transact` for short writes as you do is correct, I guess.

**Comment by careri** [comment]
Ok, sorry if I'm sounding harsh. But I'm getting quite frustrated about this. Who knows and who can explain?`

Sorry @eriohl, but I seems to recall some discussion with you and @chrhol regarding long running and performance. I had an example, can't find the issue now. That used a lot of new Transaction() which I then learned filled some pool of "fast" addressed long running transactions. I shall not assign random quotes to you anymore ;)

Someone suggested using small Db.Transact only when writing data. I like this approach, but I need to be able to read data written by other threads/transactions since it will be multi client.

@k-rus the example isn't my production code as stated, it's an example that can reproduce the problem that occurs in my real code. That's why I do thread switching.

I didn't mean that any one person contradict himself, just that the information about merging writes/transactions and Db.Scope and Db.Transact is somewhat contradicting coming from different sources

**Comment by careri** [comment]
I get the exception using Db.Scope as well, but as stated before. It works 99%
```
Singleton.RunSync(() =>
                            {
                                Db.Scope(() =>
                                {
```

**Comment by careri** [comment]
@eriohl 

> What I mean is that if you may have a Db.Transact in your callstack then Db.Transact won't create a new transaction. So it your design of rescheduling tasks on a different scheduler is because you might have a Db.Transact in your callstack then doing this won't work.

I don't follow, by switching thread I loose my callstack and therefore any Db.Transact bound to the calling thread right?`
I should therefore have full control of when the instance is written?

**Comment by eriohl** [comment]
> We've hade similar disussions before: #3192

Yes. And I kind of went out of my way not to advise you on how to write your application then as well. As I said: I generally know how it works but not how it is supposed to be used. If you ask my why something works the way it does I could probably tell you. If you ask me how you are to use this framework to write applications then I'm sorry I can't help you at all. Because I simply do not know.

> Ok, sorry if I'm sounding harsh. But I'm getting quite frustrated about this. Who knows and who can explain?

I do understand your frustration. @chrhol @dan31 @per-samuelsson: This is your stuff right? Please help @careri out.

> > What I mean is that if you may have a Db.Transact in your callstack then Db.Transact won't create a new transaction. So it your design of rescheduling tasks on a different scheduler is because you might have a Db.Transact in your callstack then doing this won't work.

> > I don't follow, by switching thread I loose my callstack and therefore any Db.Transact bound to the calling thread right?

> I should therefore have full control of when the instance is written?

Yes. What @k-rus suggested was to put a ```Db.Transact``` in the main thread **after** call to ```DoCreate```. This transaction will see the inserted instance. But this won't work if there is a previous ```Db.Transact``` on this thread (the main thread).

**Comment by eriohl** [comment]
That is:

```C#
  private Item Reload(ulong oid)
  {
     Item instance;
     Db.Transact(() => {
       instance = (Item)DbHelper.FromID(oid);
     });
     return instance;
  }
```

Now I'm not suggesting that this is good design or anything mind. But this transaction should be guaranteed to see the instance since it is guaranteed to be create **after** the transaction creating the instance.

**Comment by k-rus** [comment]
> I get the exception using Db.Scope as well, but as stated before. It works 99%
> 
> Singleton.RunSync(() =>
>                             {
>                                 Db.Scope(() =>
>                                 {

Thank you for trying. As I said I am not sure if it will work. You can also try to cover your select statement with `Db.Transact`.

I hope @chrhol can answer many of your questions.



> I like this approach, but I need to be able to read data written by other threads/transactions since it will be multi client.
> 
> @k-rus the example isn't my production code as stated, it's an example that can reproduce the problem that occurs in my real code. That's why I do thread switching.

Thus I suggest you to design your application taking in account optimistic non-locking transactions. We guarantee isolation and atomicity. But since the transactions do not use locking, the order of transactions from different threads is not defined.

**Comment by careri** [comment]
Need to go get the kids, I hope someone has a great explanation by tomorrow ;)

**Comment by k-rus** [comment]
I don't believe that the answers will really help you to solve your problem as I already tried to clarify related things. I hope it will make things more clear for you.

To really help it will be necessary to know what you need to do and probably why. For example, moving database data from one thread to another thread is not good idea, instead each thread should be designed to be independent from other threads. On the other side you might have situation that the data came to the client in response and then the client submitted another request using this data. This is valid scenario, but it involves not just transactions, but also client-server communication.

**Comment by Starcounter-Jack** [comment]
It is hard to huess the actual use case.

You need to exit your transaction and enter into a new one to get a new snapshot. For this you should use Db.Transact(). If there is a surrouding Db.Transact() higher in the call stack, you will NOT receive a new snapshot.
I  suggest you only use explicit transactions and avoid accessing the database unless in a Db.Transact for clarity and ease of reasoning.

Unless dealing with lobg running edits (such as in a UI or form), you should not use Db.Scope.

**Comment by careri** [comment]
Well, that sounds reasonable. But there is still the issue with inconsistent behaviour. My design is almost 100% working as stated above.

I would really like to know why. Each thread has a implicit readonly transaction, a DB.Transact is created if the query yielded no result. In this transact a new instance is created.

Same logic executed over and over, then suddenly exception. 
I'm distributing on different schedulers, could that be it?

I rather not rewrite again...

**Comment by k-rus** [comment]
I already answered your question. The short answer is because read-committed isolation.

You are lucky that your code does not work 100% and you notice problems of running your code in multi-threaded environment already now and was able to reproduce it.

I don't know if using the same scheduler will help and I don't recommend to rely on this.

**Comment by careri** [comment]
But if it was because of read-committed isolation, I should never get the item right?
That's whats botter me. If the outer "implicit readonly transaction/new transaction/Db.Scope" always failed to retrieve the instance commited on another thread I could accept your answer.

But most of the time the outer transaction has no problem reading the instance. I'll have a chat with @chrhol see if he can shed some light on this. My guess is that some timer is involved with updating snapshots?

**Comment by k-rus** [comment]
> But if it was because of read-committed isolation, I should never get the item right?
> That's whats botter me. If the outer "implicit readonly transaction/new transaction/Db.Scope" always failed to retrieve the instance commited on another thread I could accept your answer.

Obviously you don't understand my reply. Read committed transaction reads committed values. If value was committed it can be seen by the read committed transaction, but it does not happen immediately. The reason why it is not immediate is performance. I guess only @eriohl can explain when the committed value is visible to the read committed transaction. However, it might be changed in future.

Serialising threads as you do in your example increases the probability that the read committed transaction will see the committed value, but does not guarantee, since transactions and threads are not the same thing.

 

> My guess is that some timer is involved with updating snapshots?

I am sure, not.

**Comment by Starcounter-Jack** [comment]
You should NOT use Db.Scope or implicit transactions in this use case. Only use proper ACID transactions to get predictable results. Wrap your access into atomic Db.Transact() operations.

**Comment by Starcounter-Jack** [comment]
Db.Scope will be deprecated once branches is in place.

**Comment by careri** [comment]
Yes I understand what you say @k-rus. But this is the first time you say this:
but it does not happen immediately. 

That explains the behaviour.

I'm having a seriously hard time writing something usable that's event driven.
I want to fire of an event and send a json to my clients if e.g. persistent configuration is updated. 
I can't fire the event directly after the Db.Transact completes, because there might actually be an outer Db.Transact that pospones the actual writing to the db. So I fire of events of non existing objects.

I can't use CommitHook. Same problem as Db.Transact. The object is not written and might be rolled back.

I can't use new Transaction because of performance.

**Comment by careri** [comment]
I would like to have some way of force writing a Db.Transact no matter of outer Db.Transact. Then I wouldn't need this thread switching

**Comment by Starcounter-Jack** [comment]
> ...because there might actually be an outer Db.Transact that pospones the actual writing to the db.

If you have an outer Db.Transact, how could you expect the changes to become visible? This would violate the atomicity and isolation. Avoid any outer Db.Transact or create a ```new Transaction()``` to create an independent (non nested) transaction.

**Comment by careri** [comment]
Am I allowed to create new new Transaction within a Db.Transact? I seem the recall that's forbidden?

**Comment by Starcounter-Jack** [comment]
It should be possible to juggle independent transactions on the same thread. I'll get back to you on API...

**Comment by careri** [comment]
If I'm in a Db.Transact I get this exception when I call Commit on the newly created transaction:
> ScErrTransactionLockedOnThread (SCERR4031): Can't modify the state of the transaction.

`´´
using (var t = new Transaction(false, true))
{
    instance = new Item(m_name);
    oid = instance.GetObjectNo();
    t.Commit(); // BOOM
}
```

**Comment by eriohl** [comment]
Obviously this is a bit hard to explain. And this is on us of course. But let me try again:

a. All database access operations operates on a snapshot.
b. This snapshot is refreshed on transaction boundaries. Which is basically when a transaction is switched on the thread.
c. A transaction scope (```Db.Transact```) can be, but not always, a transaction boundary.
d. There are other kinds of transaction boundaries as well.

In your case the main thread is running within a single implicit transaction scope (starting before Main). There are no explicit transaction boundaries which in turn means that the transaction usually, but not always, runs within a single snapshot and you would not be able to see parallel transaction updates (ie snapshot isolation).

(Read-committed transactions is thus not guaranteed to see updates made by other transactions running on other threads immediately. There need to be a transaction boundary for the snapshot to be updated with the latest changes.)

There are however a lot of **potential** implicit transaction boundaries in your code. One is the thread block waiting for the parallel task of creating the new person. Another is the sleep in the Reload function. There could also be a yield to another task on the scheduler running the thread running the Main function (which can be done on timeout yes). Usually one of these implicit transaction boundaries gets triggered and your code works. The problem is that these transaction boundaries are **potential** meaning that they may not be triggered. Thus your code does not always work.

The only way you can ensure that all changes made in previous transactions is by placing an explicit transaction boundary **after** this transaction has completed. Which in your code you do not. (And how do you do this? This I can't answer unfortunately.)

**Comment by Starcounter-Jack** [comment]
The above database access is not done inside the transaction ```t```.

@chrhol is not at the office yet, but could you try:

```c#
            Transaction t = new Transaction();
            t.Scope(() => ...database access here... );
            t.Commit();
```

**Comment by k-rus** [comment]
@careri 
Thank you for explaining your use case!

I am not sure if event driven design on the server is the right approach. May be it should involve client side.

My understanding of your use case: a user submits changes on a configuration. The changes must be committed to the database and then the user view in the browser should be updated.
Do I understand the use case correctly?

I don't know what the actual mechanism is employed on the client and in the client-server communication in the case of commit. May be the user session should be closed on commit and new user session opened?

cc @alemoi @chrhol @warpech

**Comment by Starcounter-Jack** [comment]
> The only way you can ensure that all changes made in previous transactions is by placing an explicit transaction boundary **after** this transaction has completed. Which in your code you do not. (And how do you do this? This I can't answer unfortunately.)

If you create a chain of ***non-nested*** Db.Transact() calls, they will each have a transaction boundary.

```
void Foo() {
   Db.Transact( () => { ...database acess here... } ); // Will create a new snapshot and then commit.
   Db.Transact( () => { ...database acess here... } ); // Will create a new snapshot and then commit.
   Db.Transact( () => { ...database acess here... } ); // Will create a new snapshot and then commit.
}
```

**Comment by Starcounter-Jack** [comment]
@eriohl 

> ...This snapshot is refreshed on transaction boundaries. Which is basically when a transaction is switched on the thread.

Does this mean that I can have a snapshot refresh in the middle of a pending transaction? If so, how do I ensure that a snapshot never mutates during an atomic transaction?

**Comment by eriohl** [comment]
> @eriohl

> > ...This snapshot is refreshed on transaction boundaries. Which is basically when a transaction is switched on the thread.

> Does this mean that I can have a snapshot refresh in the middle of a pending transaction? If so, how do I ensure that a snapshot never mutates during an atomic transaction?

See [this](https://github.com/Starcounter/CompanyTrack/issues/162) discussion.

**Comment by careri** [comment]
@eriohl Thanks for that explenation, things are starting to fall into place now. I didn't understand why you suggested to use a Db.Transact in the Reload-method above. But it's because of the transaction boundry.

@k-rus The clients are not web-based. I have a administrator server where you configure what software shall be installed on which computer.

The admin UI is web-based. Through the admin UI you can start an install/upgrade of a software. This pushes out messages to the clients to start installing. The clients report the progress to the server. The server aggregates the progress and displays it in the Admin UI.

**Comment by careri** [comment]
@Starcounter-Jack 
> BTW, the Db.Transact() function is just thin sugaring on top of the COMMIT semantics using a nesting counter (same as BEGIN TRANSACTION END TRANSACTION in T-SQL).

Is this true? Someone,probably @chrhol told me this:
Using ```new Transaction```will always put it in the pool of long running transactions, the same goes for Db.Scope.

There is a limited slots for fast access of transactions in this pool. This was an issue I reported long ago that contained a lot of nested ```new Transaction```.

I then switched to Db.Transact, which was explain to me as a light weight transaction. Some type of handle directly into the kernel that was way faster than using Transaction.

Is this information correct?

**Comment by Starcounter-Jack** [comment]
@careri 
I need to check on this. ```Db.Transact()``` was designed with the same goal as ```BEGIN/END```, namely to help the common nesting pattern. I.e. to protect an outer transaction to avoid inner transactions to leak.

I need to check on how to juggle transactions without ```Db.Transact()```. From the standpoint of the database engine, this should not be a problem. The ````Db.Transact()``` function is purely a C# concept. The underlying API treats transactions as entities that can be freely created and used.

As soon as @chrhol comes to the office, I will check with him.

**Comment by careri** [comment]
There seems to be no way using a Transaction in a Transact.
```
Item item = null;

Db.Transact(() =>
    {
        try
        {
            using (var t = new Transaction(false))
            {
                item = new Item("fail");
                t.Commit(); // BOOM
            }
        }
        catch { }
        try
        {
            using (var t2 = new Transaction(false))
            {
                item = t2.Scope(() => new Item("fail")); // BOOM
                t2.Commit();
            }
        }
        catch { }
    });
```

**Comment by Starcounter-Jack** [comment]
@careri 

This is flawed. We will put priority on this. You must be able to create transactions manually.

**Comment by k-rus** [comment]
IMHO fixing transactions to work as requested in this case is wrong approach. Furthermore, the use case can be designed to work with the current Starcounter. However, it is difficult for me to advise how the desired functionality of the application should be implemented as I don't know much about client-server interactions in terms of transactions.

**Comment by Starcounter-Jack** [comment]
@k-rus 

The C# API is kind of flawed. We will discuss the API today is you are interested to join.

**Comment by careri** [comment]
That would simplify things for me. We had that feature in scdb1x:
```
[TransactionScope(TransactionScopeOptions.RequiresNew)]
       private PostAddressComponent DoApprove(NewPostalAddress adr)
``´

**Comment by careri** [comment]
I can confirm that using a NOOP Db.Transact before leaving the DoCreate-method does update the snapshot of the implicit readonly transaction.

I'll use this ugly solution for now. Thanks for all the input and information.

@k-rus you're probably right about me being able to design it better. But then I would need to keep track of modified/created instances in some other fashion. But this is where I ended up after being pointed in new directions every time I reported an issue.

This could be done with polling the database for changes of monitored instances. Or by using a transaction log reader.
We'll see if I'll run into any performance issues with this solution, I doubt it since reading already existing instances is done in the same Db.Transact. Only creating new instances with unique index constraint are created in a thread switch.

Looking forward to  the outcome of you're discussions.
